File: Android/VeLiveQuickStartDemo/app/src/main/java/com/ttsdk/quickstart/features/interact/manager/VeLiveAudienceManager.java
/**
 * Copyright (c) 2023 Beijing Volcano Engine Technology Ltd. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */
package com.ttsdk.quickstart.features.interact.manager;


import static com.ss.bytertc.engine.VideoCanvas.RENDER_MODE_HIDDEN;
import static com.ss.videoarch.liveplayer.VeLivePlayerDef.VeLivePlayerFillMode.VeLivePlayerFillModeAspectFill;
import static com.ss.videoarch.liveplayer.VeLivePlayerDef.VeLivePlayerStatus.VeLivePlayerStatusError;
import static com.ss.videoarch.liveplayer.VeLivePlayerDef.VeLivePlayerStatus.VeLivePlayerStatusPaused;
import static com.ss.videoarch.liveplayer.VeLivePlayerDef.VeLivePlayerStatus.VeLivePlayerStatusPlaying;
import static com.ss.videoarch.liveplayer.VeLivePlayerDef.VeLivePlayerStatus.VeLivePlayerStatusPrepared;
import static com.ss.videoarch.liveplayer.VeLivePlayerDef.VeLivePlayerStatus.VeLivePlayerStatusStopped;

import android.graphics.Bitmap;
import android.text.TextUtils;
import android.util.Log;
import android.view.SurfaceHolder;
import android.view.TextureView;

import com.pandora.common.env.Env;
import com.ss.bytertc.engine.RTCRoom;
import com.ss.bytertc.engine.RTCRoomConfig;
import com.ss.bytertc.engine.RTCVideo;
import com.ss.bytertc.engine.UserInfo;
import com.ss.bytertc.engine.VideoCanvas;
import com.ss.bytertc.engine.VideoEncoderConfig;
import com.ss.bytertc.engine.data.MirrorType;
import com.ss.bytertc.engine.data.StreamIndex;
import com.ss.bytertc.engine.handler.IRTCVideoEventHandler;
import com.ss.bytertc.engine.type.ChannelProfile;
import com.ss.bytertc.engine.type.MediaStreamType;
import com.ss.bytertc.engine.type.RTCRoomStats;
import com.ss.bytertc.engine.type.StreamRemoveReason;
import com.ss.bytertc.engine.video.VideoCaptureConfig;
import com.ss.videoarch.liveplayer.VeLivePlayer;
import com.ss.videoarch.liveplayer.VeLivePlayerAudioFrame;
import com.ss.videoarch.liveplayer.VeLivePlayerConfiguration;
import com.ss.videoarch.liveplayer.VeLivePlayerDef;
import com.ss.videoarch.liveplayer.VeLivePlayerError;
import com.ss.videoarch.liveplayer.VeLivePlayerObserver;
import com.ss.videoarch.liveplayer.VeLivePlayerStatistics;
import com.ss.videoarch.liveplayer.VeLivePlayerVideoFrame;
import com.ss.videoarch.liveplayer.VideoLiveManager;

public class VeLiveAudienceManager {
    private static final String TAG = "VeLiveAudienceManager";

    private IListener mRoomListener;
    private static VeLiveAudienceManager mInstance;
    private final String mAppId;
    private String mUserId;
    private String mPullUrl;
    private VeLiveAudienceManager.Config mConfig;

    private VeLivePlayer mLivePlayer;                             //  播放器引擎  

    private RTCVideo mRTCVideo;
    private RTCRoom mRTCRoom; //  RTC 房间对象  
    private String mRoomId;
    private TextureView mLocalView;
    private int streamRetryCount;

    public static class Config {
        public int mVideoCaptureWidth = 720;
        public int mVideoCaptureHeight = 1280;
        public int mVideoCaptureFps = 30;

        public int mAudioCaptureSampleRate = 44100;
        public int mAudioCaptureChannel = 2;

        public int mVideoEncoderWidth = 720;
        public int mVideoEncoderHeight = 1280;
        public int mVideoEncoderFps = 15;
        public int mVideoEncoderKBitrate = 1600;
        public boolean mIsVideoHardwareEncoder = true;

        public int mAudioEncoderSampleRate = 44100;
        public int mAudioEncoderChannel = 2;
        public int mAudioEncoderKBitrate = 32;
    }

    public interface IListener {
        void onUserJoined(String uid);
        void onUserLeave(String uid);
        void onJoinRoom(int state);
        void onUserPublishStream(String uid, MediaStreamType type);
        void onUserUnPublishStream(String uid, MediaStreamType type, StreamRemoveReason reason);
    }

    private VeLiveAudienceManager(String appId, String userId) {
        mAppId = appId;
        mUserId = userId;
        initLivePlayer();
    }

    public static VeLiveAudienceManager create(String appId, String userId) {
        if (mInstance == null) {
            if (!TextUtils.isEmpty(appId) && !TextUtils.isEmpty(userId)) {
                mInstance = new VeLiveAudienceManager(appId, userId);
            }
        }
        return mInstance;
    }

    public static void destroy() {
        if (mInstance != null) {
            mInstance.stopInteract();
            mInstance.stopPlay();

            mInstance.releaseLivePlayer();
        }
        mInstance = null;
    }

    public RTCVideo getRTCVideo() {
        return mRTCVideo;
    }
    public void setConfig(VeLiveAudienceManager.Config config) {
        mConfig = config;
    }

    public void setPlayerVideoView(SurfaceHolder holder) {

        //  设置view  
        mLivePlayer.setSurfaceHolder(holder);
        //  设置渲染模式  
        mLivePlayer.setRenderFillMode(VeLivePlayerFillModeAspectFill);
    }

    public void startPlay(String url) {
        if (mLivePlayer == null) {
            return;
        }
        streamRetryCount = 0;
        mPullUrl = url;
        mLivePlayer.setPlayUrl(mPullUrl);
        mLivePlayer.play();
    }

    public void stopPlay() {
        if (mLivePlayer == null) {
            return;
        }
        mLivePlayer.stop();
    }

    public void setLocalVideoView(TextureView renderView) {
        mLocalView = renderView;
        if (mRTCVideo == null) {
            return;
        }
        VideoCanvas videoCanvas = new VideoCanvas();
        videoCanvas.renderView = renderView;
        videoCanvas.uid = mUserId;
        videoCanvas.isScreen = false;
        videoCanvas.renderMode = VideoCanvas.RENDER_MODE_HIDDEN;
        //  设置本地视频渲染视图  
        mRTCVideo.setLocalVideoCanvas(StreamIndex.STREAM_INDEX_MAIN, videoCanvas);
        Log.d(TAG, "setLocalVideoView");
    }

    public void setRemoteVideoView(String uid, TextureView renderView) {
        if (mRTCVideo != null) {
            VideoCanvas canvas = new VideoCanvas(renderView, RENDER_MODE_HIDDEN, mRoomId, uid, false);
            mRTCVideo.setRemoteVideoCanvas(uid, StreamIndex.STREAM_INDEX_MAIN, canvas);
        }
    }

    public void startAudioCapture() {
        if (mRTCVideo != null) {
            mRTCVideo.startAudioCapture();
        }
    }

    public void startVideoCapture() {
        if (mRTCVideo != null) {
            VideoEncoderConfig config = new VideoEncoderConfig(
                    mConfig.mVideoEncoderWidth,
                    mConfig.mVideoEncoderHeight,
                    mConfig.mVideoEncoderFps,
                    mConfig.mVideoEncoderKBitrate * 1000);
            mRTCVideo.setVideoEncoderConfig(config);
            mRTCVideo.startVideoCapture();
        }
    }

    public void stopAudioCapture() {
        if (mRTCVideo != null) {
            mRTCVideo.stopAudioCapture();
        }
    }

    public void stopVideoCapture() {
        if (mRTCVideo != null) {
            mRTCVideo.stopVideoCapture();
        }
    }

    public void startInteract(String roomId, String token, VeLiveAudienceManager.IListener roomListener) {
        stopPlay();
        initRtcEngine();
        startAudioCapture();
        startVideoCapture();
        setLocalVideoView(mLocalView);
        mRoomListener = roomListener;
        //  设置用户信息，加入房间，开始连麦  
        joinRoom(roomId, mUserId, token, mRoomListener);
    }

    public void stopInteract() {
        leaveRoom();
        stopVideoCapture();
        stopAudioCapture();
        setLocalVideoView(null);
        startPlay(mPullUrl);
        releaseRtcEngine();
    }

    private void initLivePlayer() {
        //  创建播放器  
        mLivePlayer = new VideoLiveManager(Env.getApplicationContext());

        //  设置播放器回调  
        mLivePlayer.setObserver(mLivePlayerObserver);

        //  播放器基础设置  
        VeLivePlayerConfiguration config = new VeLivePlayerConfiguration();
        config.enableStatisticsCallback = true;
        config.enableLiveDNS = true;
        mLivePlayer.setConfig(config);
    }

    public void sendSeiMessage(String msg, int repeat) {
        mRTCVideo.sendSEIMessage(StreamIndex.STREAM_INDEX_MAIN,
                TextUtils.isEmpty(msg) ? new byte[0] : msg.getBytes(), repeat);
    }

    private void releaseLivePlayer() {
        if (mLivePlayer != null) {
            mLivePlayer.destroy();
            mLivePlayer = null;
        }
        if (mLocalView != null) {
            mLocalView = null;
        }
    }

    private final IRTCVideoEventHandler mRTCVideoEventHandler = new IRTCVideoEventHandler() {
        @Override
        public void onWarning(int warn) {
            super.onWarning(warn);
            Log.i(TAG, warn + "");
        }

        @Override
        public void onError(int err) {
            super.onError(err);
            Log.i(TAG, err + "");
        }

        @Override
        public void onNetworkTypeChanged(int type) {
            super.onNetworkTypeChanged(type);
        }
    };

    private void initRtcEngine() {
        mRTCVideo = RTCVideo.createRTCVideo(Env.getApplicationContext(), mAppId, mRTCVideoEventHandler, null, null);
        mRTCVideo.setLocalVideoMirrorType(MirrorType.MIRROR_TYPE_RENDER_AND_ENCODER);
        VideoCaptureConfig captureConfig = new VideoCaptureConfig();
        captureConfig.width = mConfig.mVideoCaptureWidth;
        captureConfig.height = mConfig.mVideoCaptureHeight;
        captureConfig.frameRate = mConfig.mVideoCaptureFps;
        mRTCVideo.setVideoCaptureConfig(captureConfig);
    }

    private void releaseRtcEngine() {
        if (mRTCRoom != null) {
            mRTCRoom.destroy();
            mRTCRoom = null;
        }
        if (mRTCVideo != null) {
            RTCVideo.destroyRTCVideo();
            mRTCVideo = null;
        }
    }

    private RtcRoomEventHandlerAdapter mIRtcRoomEventHandler = new RtcRoomEventHandlerAdapter() {
        @Override
        public void onLeaveRoom(RTCRoomStats stats) {
        }

        @Override
        public void onRoomStateChanged(String roomId, String uid, int state, String extraInfo) {
            if (TextUtils.equals(uid, mUserId) && TextUtils.equals(roomId, mRoomId)) {
                mRoomListener.onJoinRoom(state);
            }
        }

        @Override
        public void onUserJoined(UserInfo userInfo, int elapsed) {
            mRoomListener.onUserJoined(userInfo.getUid());
        }

        @Override
        public void onUserLeave(String uid, int reason) {
            mRoomListener.onUserLeave(uid);
        }

        @Override
        public void onUserPublishStream(String uid, MediaStreamType type) {
            mRoomListener.onUserPublishStream(uid, type);
        }

        @Override
        public void onUserUnpublishStream(String uid, MediaStreamType type, StreamRemoveReason reason) {
            mRoomListener.onUserUnPublishStream(uid, type, reason);
        }
    };

    private void joinRoom(String roomId, String userId, String token, VeLiveAudienceManager.IListener roomListener) {
        Log.d(TAG, String.format("joinRoom: %s %s %s", roomId, userId, token));
        mRoomListener = roomListener;
        if (mRTCVideo == null) {
            return;
        }
        mRTCRoom = mRTCVideo.createRTCRoom(roomId);
        mRTCRoom.setRTCRoomEventHandler(mIRtcRoomEventHandler);
        mUserId = userId;
        mRoomId = roomId;
        UserInfo userInfo = new UserInfo(userId, null);
        RTCRoomConfig roomConfig = new RTCRoomConfig(ChannelProfile.CHANNEL_PROFILE_COMMUNICATION,
                true, true, true);
        mRTCRoom.joinRoom(token, userInfo, roomConfig);
    }

    private void leaveRoom() {
        Log.d(TAG, "leaveRoom");
        if (mRTCRoom != null) {
            mRTCRoom.leaveRoom();
            mRTCRoom = null;
        }
    }

    private VeLivePlayerObserver mLivePlayerObserver = new VeLivePlayerObserver() {
        @Override
        public void onError(VeLivePlayer player, VeLivePlayerError error) {
            Log.i("VeLivePlayer-", "Player Error: " + error.mErrorCode);
        }

        @Override
        public void onFirstVideoFrameRender(VeLivePlayer player, boolean isFirstFrame) {
            Log.i("VeLivePlayer-", "First Video Frame: " + isFirstFrame);
        }

        @Override
        public void onFirstAudioFrameRender(VeLivePlayer player, boolean isFirstFrame) {
            Log.i("VeLivePlayer-", "First Audio Frame: " + isFirstFrame);
        }

        @Override
        public void onStallStart(VeLivePlayer player) {
            Log.i("VeLivePlayer-", "Stall Start");
        }

        @Override
        public void onStallEnd(VeLivePlayer player) {
            Log.i("VeLivePlayer-", "Stall End");
        }

        @Override
        public void onVideoRenderStall(VeLivePlayer player, long stallTime) {
            Log.i("VeLivePlayer-", "Video Render Stall：" + stallTime);
        }

        @Override
        public void onAudioRenderStall(VeLivePlayer player, long stallTime) {
            Log.i("VeLivePlayer-", "Audio Render Stall：" + stallTime);
        }

        @Override
        public void onResolutionSwitch(VeLivePlayer player, VeLivePlayerDef.VeLivePlayerResolution resolution, VeLivePlayerError error, VeLivePlayerDef.VeLivePlayerResolutionSwitchReason reason) {
            Log.i("VeLivePlayer-", "Resolution Switch：" + resolution);
        }

        @Override
        public void onVideoSizeChanged(VeLivePlayer player, int width, int height) {
            Log.i("VeLivePlayer-", "Video Size Changed width:" + width + " height:" + height);
        }

        @Override
        public void onReceiveSeiMessage(VeLivePlayer player, String message) {
            Log.i("VeLivePlayer-", "SEI:" + message);
        }

        @Override
        public void onMainBackupSwitch(VeLivePlayer player, VeLivePlayerDef.VeLivePlayerStreamType streamType, VeLivePlayerError error) {
            Log.i("VeLivePlayer-", "Main Backup Switch");
        }

        @Override
        public void onPlayerStatusUpdate(VeLivePlayer player, VeLivePlayerDef.VeLivePlayerStatus status) {
            if(status == VeLivePlayerStatusPrepared) {
                Log.i("VeLivePlayer-", "State:" + "Prepared");
            }else if(status == VeLivePlayerStatusPlaying) {
                Log.i("VeLivePlayer-", "State:" + "Playing");
            }else if(status == VeLivePlayerStatusPaused) {
                Log.i("VeLivePlayer-", "State:" + "Paused");
            }else if(status == VeLivePlayerStatusStopped) {
                Log.i("VeLivePlayer-", "State:" + "Stopped");
            }else if(status == VeLivePlayerStatusError) {
                Log.i("VeLivePlayer-", "State:" + "Error");
            }
        }

        @Override
        public void onStatistics(VeLivePlayer player, VeLivePlayerStatistics statistics) {
            Log.i("VeLivePlayer-", "statistics url:" + statistics.url);
            Log.i("VeLivePlayer-", "statistics width:" + statistics.width);
            Log.i("VeLivePlayer-", "statistics height:" + statistics.height);
            Log.i("VeLivePlayer-", "statistics isHardWareDecode:" + statistics.isHardwareDecode);
            Log.i("VeLivePlayer-", "statistics videoCodec:" + statistics.videoCodec);
            Log.i("VeLivePlayer-", "statistics bitrate:" + statistics.bitrate);
            Log.i("VeLivePlayer-", "statistics fps:" + statistics.fps);
        }

        @Override
        public void onSnapshotComplete(VeLivePlayer player, Bitmap bitmap) {
            Log.i("VeLivePlayer-", "Snapshot");
        }

        @Override
        public void onRenderVideoFrame(VeLivePlayer player, VeLivePlayerVideoFrame videoFrame) {
            Log.i("VeLivePlayer-", "Render Video Frame");
        }

        @Override
        public void onRenderAudioFrame(VeLivePlayer player, VeLivePlayerAudioFrame audioFrame) {
            Log.i("VeLivePlayer-", "Render Audio Frame");
        }

        @Override
        public void onStreamFailedOpenSuperResolution(VeLivePlayer veLivePlayer, VeLivePlayerError veLivePlayerError) {

        }
    };
}

File: iOS/VeLiveQuickStartDemo/VeLiveQuickStartDemo/Features/Interact/Manager/VeLiveAudienceManager.m
/**
 * Copyright (c) 2023 Beijing Volcano Engine Technology Ltd. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */

//
//  VeLiveAudienceManager.m
//  VeLiveQuickStartDemo
//
//  Created by Volcano Engine Team on 2023/2/22.
//

#import "VeLiveAudienceManager.h"
@interface VeLiveAudienceManager () <ByteRTCRoomDelegate, ByteRTCVideoDelegate, VeLivePlayerObserver>
@property (nonatomic, strong, readwrite) ByteRTCVideo *rtcVideo;
@property (nonatomic, strong, readwrite) ByteRTCRoom *rtcRoom;
@property (nonatomic, copy, readwrite) NSString *appId;
@property (nonatomic, copy, readwrite) NSString *userId;
@property (nonatomic, copy, readwrite) NSString *roomId;
@property (nonatomic, copy, readwrite) NSString *token;
@property (nonatomic, assign, getter=isInteractive) BOOL interactive;
@property (nonatomic, copy) NSString *playUrl;
@property (nonatomic, strong, readwrite) TVLManager *livePlayer;
@property (nonatomic, assign) NSInteger streamRetryCount;
@end
@implementation VeLiveAudienceManager
- (instancetype)initWithAppId:(NSString *)appId userId:(NSString *)userId {
    if (self = [super init]) {
        self.appId = appId;
        self.userId = userId;
        self.config = [[VeLiveConfig alloc] init];
    }
    return self;
}

- (instancetype)init {
    return [self initWithAppId:@"" userId:@""];
}

- (void)setLocalVideoView:(UIView *)localVideoView {
    _localVideoView = localVideoView;
    [self setupLocalVideoView:localVideoView];
}

- (void)setRemoteVideoView:(UIView *)view forUid:(NSString *)uid {
    if (self.rtcVideo) {
        ByteRTCVideoCanvas *canvas = [[ByteRTCVideoCanvas alloc] init];
        canvas.renderMode = ByteRTCRenderModeHidden;
        canvas.view.backgroundColor = [UIColor clearColor];
        canvas.view = view;
        
        ByteRTCRemoteStreamKey *streamKey = [[ByteRTCRemoteStreamKey alloc] init];
        streamKey.userId = uid;
        streamKey.streamIndex = ByteRTCStreamIndexMain;
        streamKey.roomId = self.roomId;
        [self.rtcVideo setRemoteVideoCanvas:streamKey withCanvas:canvas];
    }
}

- (void)startVideoCapture {
    [self createRTCVideoIfNeed];
    __weak __typeof__(self)weakSelf = self;
    [self authorPermissionFor:(AVMediaTypeVideo) completion:^(BOOL granted) {
        __strong __typeof__(weakSelf)self = weakSelf;
        if (granted) {
            [self.rtcVideo startVideoCapture];
        } else {
            NSLog(@"VeLiveQuickStartDemo: Please turn on the camera permission");
        }
    }];
    
}

- (void)stopVideoCapture {
    [self.rtcVideo stopVideoCapture];
    [self setupLocalVideoView:nil];
}

- (void)startAudioCapture {
    [self createRTCVideoIfNeed];
    __weak __typeof__(self)weakSelf = self;
    [self authorPermissionFor:(AVMediaTypeAudio) completion:^(BOOL granted) {
        __strong __typeof__(weakSelf)self = weakSelf;
        if (granted) {
            [self.rtcVideo startAudioCapture];
        } else {
            NSLog(@"VeLiveQuickStartDemo: Please turn on the microphone permission");
        }
    }];
    
}

- (void)stopAudioCapture {
    [self.rtcVideo stopAudioCapture];
}

- (void)startInteract:(NSString *)roomId token:(NSString *)token delegate:(id <VeLiveAudienceDelegate>)delegate {
    [self stopPlay];
    [self startAudioCapture];
    [self startVideoCapture];
    
    self.roomId = roomId;
    self.token = token;
    self.delegate = delegate;
    [self setupRTCRoomIfNeed];
    
    //  设置用户信息  
    ByteRTCUserInfo *userInfo = [[ByteRTCUserInfo alloc] init];
    userInfo.userId = self.userId;
    
    //  加入房间，开始连麦  
    ByteRTCRoomConfig *config = [ByteRTCRoomConfig new];
    config.isAutoPublish = YES;
    config.isAutoSubscribeAudio = YES;
    config.isAutoSubscribeVideo = YES;
    
    //  加入RTC房间  
    NSLog(@"VeLiveQuickStartDemo: join room %@ - %@", roomId, self.userId);
    [self.rtcRoom joinRoom:token userInfo:userInfo roomConfig:config];
    self.interactive = YES;
}


- (void)stopInteract {
    [self.rtcRoom leaveRoom];
    self.interactive = NO;
    [self stopAudioCapture];
    [self stopVideoCapture];
    [self startPlayIfNeed];
}

- (void)sendSeiMessage:(NSString *)message repeat:(int)repeat {
    [self.rtcVideo sendSEIMessage:(ByteRTCStreamIndexMain)
                       andMessage:[message dataUsingEncoding:NSUTF8StringEncoding]
                   andRepeatCount:repeat
                 andCountPerFrame:(kSingleSEIPerFrame)];
}

- (void)destory {
    [self stopPlay];
    [self.livePlayer destroy];
    self.livePlayer = nil;
    [self stopInteract];
    [self stopVideoCapture];
    [self stopAudioCapture];
    [self.rtcRoom destroy];
    self.rtcRoom = nil;
    [ByteRTCVideo destroyRTCVideo];
    self.rtcVideo = nil;
}

// MARK: - ByteRTCVideoDelegate
- (void)rtcEngine:(ByteRTCVideo *_Nonnull)engine onWarning:(ByteRTCWarningCode)Code {
    NSLog(@"VeLiveQuickStartDemo: rtc on Warning");
}

- (void)rtcEngine:(ByteRTCVideo *_Nonnull)engine onError:(ByteRTCErrorCode)errorCode {
    NSLog(@"VeLiveQuickStartDemo: rtc on Error");
}

- (void)rtcEngine:(ByteRTCVideo *_Nonnull)engine onCreateRoomStateChanged:(NSString * _Nonnull)roomId errorCode:(NSInteger)errorCode {
    if (errorCode != 0) {
        NSLog(@"VeLiveQuickStartDemo: on create room error");
    }
}

// MARK: - ByteRTCRoomDelegate
- (void)rtcRoom:(ByteRTCRoom *_Nonnull)rtcRoom
   onRoomStateChanged:(NSString *_Nonnull)roomId
            withUid:(nonnull NSString *)uid
          state:(NSInteger)state
      extraInfo:(NSString *_Nonnull)extraInfo {
    dispatch_async(dispatch_get_main_queue(), ^{
        if (state == 0) {
            if (self.delegate && [self.delegate respondsToSelector:@selector(manager:onJoinRoom:)]) {
                [self.delegate manager:self onJoinRoom:uid];
            }
        } else {
            NSLog(@"VeLiveQuickStartDemo: on join room error");
        }
    });
}

- (void)rtcRoom:(ByteRTCRoom *)rtcRoom onUserJoined:(ByteRTCUserInfo *)userInfo elapsed:(NSInteger)elapsed {
    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.delegate && [self.delegate respondsToSelector:@selector(manager:onUserJoined:)]) {
            [self.delegate manager:self onUserJoined:userInfo.userId];
        }
    });
}

- (void)rtcRoom:(ByteRTCRoom *)rtcRoom onUserLeave:(NSString *)uid reason:(ByteRTCUserOfflineReason)reason {
    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.delegate && [self.delegate respondsToSelector:@selector(manager:onUserLeave:)]) {
            [self.delegate manager:self onUserLeave:uid];
        }
    });
}

- (void)rtcRoom:(ByteRTCRoom *)rtcRoom onUserPublishStream:(NSString *)userId type:(ByteRTCMediaStreamType)type {
    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.delegate && [self.delegate respondsToSelector:@selector(manager:onUserPublishStream:type:)]) {
            [self.delegate manager:self onUserPublishStream:userId type:type];
        }
    });
}

- (void)rtcRoom:(ByteRTCRoom *)rtcRoom onUserUnpublishStream:(NSString * _Nonnull)userId type:(ByteRTCMediaStreamType)type reason:(ByteRTCStreamRemoveReason)reason {
    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.delegate && [self.delegate respondsToSelector:@selector(manager:onUserUnPublishStream:type:reason:)]) {
            [self.delegate manager:self onUserUnPublishStream:userId type:type reason:reason];
        }
    });
}

// MARK: - Private
- (void)createRTCVideoIfNeed {
    if (self.rtcVideo == nil && self.appId != nil && self.appId.length > 0) {
        self.rtcVideo = [ByteRTCVideo createRTCVideo:self.appId delegate:self parameters:@{}];
        ByteRTCVideoEncoderConfig *solution = [[ByteRTCVideoEncoderConfig alloc] init];
        solution.width = self.config.captureWidth;
        solution.height = self.config.captureHeight;
        solution.frameRate = self.config.captureFps;
        solution.maxBitrate = self.config.videoEncoderKBitrate;
        [self.rtcVideo setVideoOrientation:(ByteRTCVideoOrientationPortrait)];
        [self.rtcVideo setMaxVideoEncoderConfig:solution];
        [self setupLocalVideoView:_localVideoView];
    }
}

- (void)setupLocalVideoView:(UIView *)view {
    if (_rtcVideo) {
        //  设置本地View  
        ByteRTCVideoCanvas *canvasView = [[ByteRTCVideoCanvas alloc] init];
        canvasView.view = view;
        canvasView.renderMode = ByteRTCRenderModeHidden;
        [self.rtcVideo setLocalVideoCanvas:ByteRTCStreamIndexMain withCanvas:canvasView];
    }
    
    if (_livePlayer) {
        [view addSubview:_livePlayer.playerView];
        _livePlayer.playerView.frame = view.bounds;
    }
}

- (void)setupRTCRoomIfNeed {
    //  创建 RTC 房间  
    if (self.rtcRoom == nil) {
        self.rtcRoom = [self.rtcVideo createRTCRoom:self.roomId];
        self.rtcRoom.delegate = self;
    }
}

- (void)authorPermissionFor:(AVMediaType)type completion:(void (^)(BOOL granted))completion  {
    AVAuthorizationStatus status = [AVCaptureDevice authorizationStatusForMediaType:type];
    if (status == AVAuthorizationStatusNotDetermined || status == AVAuthorizationStatusRestricted) {
        [AVCaptureDevice requestAccessForMediaType:type completionHandler:^(BOOL granted) {
            dispatch_async(dispatch_get_main_queue(), ^{
                completion(granted);
            });
        }];
    } else {
        completion(status == AVAuthorizationStatusAuthorized);
    }
}

- (void)dealloc {
    [self destory];
}

- (void)startPlay:(NSString *)url {
    [self setupPlayerIfNeed];
    self.playUrl = url;
    [self.livePlayer setPlayUrl:url];
    [self.livePlayer play];
    self.livePlayer.playerView.hidden = NO;
    [self setupLocalVideoView:self.localVideoView];
}

- (void)stopPlay {
    [NSObject cancelPreviousPerformRequestsWithTarget:self];
    self.streamRetryCount = 0;
    [self.livePlayer stop];
    self.livePlayer.playerView.hidden = YES;
}

// MARK: - Private
- (void)startPlayIfNeed {
    if (self.livePlayer) {
        [self startPlay:self.playUrl];
    }
}
- (void)setupPlayerIfNeed {
    if (self.livePlayer == nil) {
        //  创建播放器  
        TVLManager *livePlayer =  [[TVLManager alloc] initWithOwnPlayer:YES];
        self.livePlayer = livePlayer;
        //  设置播放器回调  
        [self.livePlayer setObserver:self];
        //  播放器基础设置  
        VeLivePlayerConfiguration *config = [[VeLivePlayerConfiguration alloc]init];
        config.enableStatisticsCallback = YES;
        config.enableLiveDNS = YES;
        [self.livePlayer setConfig:config];
    }
    
    if (self.livePlayer.playerView.superview != self.localVideoView) {
        [self.livePlayer.playerView removeFromSuperview];
        [self.localVideoView addSubview:self.livePlayer.playerView];
        self.livePlayer.playerView.frame = self.localVideoView.bounds;
    }
    
    //  设置渲染模式  
    [self.livePlayer setRenderFillMode:VeLivePlayerFillModeAspectFill];
}

#pragma mark - VeLivePlayerObserver

- (void)onError:(TVLManager *)player error:(VeLivePlayerError *)error {
    NSLog(@"VeLiveQuickStartDemo: Play Error: %ld", error.errorCode);
}

- (void)onFirstVideoFrameRender:(TVLManager *)player isFirstFrame:(BOOL)isFirstFrame {
    NSLog(@"VeLiveQuickStartDemo: First Video Frame: %d", isFirstFrame);
}

- (void)onFirstAudioFrameRender:(TVLManager *)player isFirstFrame:(BOOL)isFirstFrame {
    NSLog(@"VeLiveQuickStartDemo: First Audio Frame: %d", isFirstFrame);
}

- (void)onStallStart:(TVLManager *)player {
    NSLog(@"VeLiveQuickStartDemo: Stall Start");
}

- (void)onStallEnd:(TVLManager *)player {
    NSLog(@"VeLiveQuickStartDemo: Stall End");
}

- (void)onVideoRenderStall:(TVLManager *)player stallTime:(int64_t)stallTime  {
    NSLog(@"VeLiveQuickStartDemo: Video Render Stall：%lld", stallTime);
}

- (void)onAudioRenderStall:(TVLManager *)player stallTime:(int64_t)stallTime  {
    NSLog(@"VeLiveQuickStartDemo: Video Render Stall：%lld", stallTime);
}

- (void)onResolutionSwitch:(TVLManager *)player resolution:(VeLivePlayerResolution)resolution error:(VeLivePlayerError *)error reason:(VeLivePlayerResolutionSwitchReason)reason  {
    NSLog(@"VeLiveQuickStartDemo: Resolution Switch resolution=%ld", resolution);
}

- (void)onVideoSizeChanged:(TVLManager *)player width:(int)width height:(int)height  {
    NSLog(@"VeLiveQuickStartDemo: Video Size Change width=%d height=%d", width, height);
}

- (void)onReceiveSeiMessage:(TVLManager *)player message:(NSString*)message  {
    NSLog(@"VeLiveQuickStartDemo: SEI :%@",message);
}

- (void)onMainBackupSwitch:(TVLManager *)player streamType:(VeLivePlayerStreamType)streamType error:(VeLivePlayerError *)error  {
    NSLog(@"VeLiveQuickStartDemo: Main Backup Switch");
}

- (void)onPlayerStatusUpdate:(TVLManager *)player status:(VeLivePlayerStatus)status  {
    if(status == VeLivePlayerStatusPrepared) {
        NSLog(@"VeLiveQuickStartDemo: State: Prepared");
    }else if(status == VeLivePlayerStatusPlaying) {
        NSLog(@"VeLiveQuickStartDemo: State: Playing");
    }else if(status == VeLivePlayerStatusPaused) {
        NSLog(@"VeLiveQuickStartDemo: State: Paused");
    }else if(status == VeLivePlayerStatusStopped) {
        NSLog(@"VeLiveQuickStartDemo: State: Stopped");
    }else if(status == VeLivePlayerStatusError) {
        NSLog(@"VeLiveQuickStartDemo: State: Error");
    }
}

- (void)onStatistics:(TVLManager *)player statistics:(VeLivePlayerStatistics *)statistics {
    NSLog(@"VeLiveQuickStartDemo: statistics.url=%@", statistics.url);
    NSLog(@"VeLiveQuickStartDemo: statistics.width=%d", statistics.width);
    NSLog(@"VeLiveQuickStartDemo: statistics.height=%d", statistics.height);
    NSLog(@"VeLiveQuickStartDemo: statistics.isHardWareDecode=%d", statistics.isHardWareDecode);
    NSLog(@"VeLiveQuickStartDemo: statistics.videoCodec=%@", statistics.videoCodec);
    NSLog(@"VeLiveQuickStartDemo: statistics.bitrate=%ld", statistics.bitrate);
    NSLog(@"VeLiveQuickStartDemo: statistics.fps=%f", statistics.fps);
}

- (void)onSnapshotComplete:(TVLManager *)player image:(UIImage *)image  {
    NSLog(@"VeLiveQuickStartDemo: snapshot");
}

- (void)onRenderVideoFrame:(TVLManager *)player videoFrame:(VeLivePlayerVideoFrame *)videoFrame  {
    NSLog(@"VeLiveQuickStartDemo: on render video frame");
}

- (void)onRenderAudioFrame:(TVLManager *)player audioFrame:(VeLivePlayerAudioFrame *)audioFrame  {
    NSLog(@"VeLiveQuickStartDemo: on render audio frame");
}


@end
